clippy_lints/src/volatile_composites.rs

```diff
@@ -0,0 +1,203 @@
+}
+declare_lint_pass!(VolatileComposites => [VOLATILE_COMPOSITES]);
+
+// functions:
+// core::ptr::{read_volatile,write_volatile}
+// methods:
+// pointer::{read_volatile,write_volatile}
+// NonNull::{read_volatile,write_volatile}
+
+// primitive type:
+// unit, [iu]{8,16,32,64,128?}, f{32,64}, thin pointer, usize, isize, bool, char
+// C enum with primitive repr
+// #[repr(transparent)] wrapper of above
+
+// Zero-sized types are intrinsically safe to use volatile on since they won't
+// actually generate *any* loads or stores. But this is also used to skip zero
+// fields of #[repr(transparent)] structures.
+fn is_zero_sized_ty<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    if let Ok(ty) = cx.tcx.try_normalize_erasing_regions(cx.typing_env(), ty)
+        && let Ok(layout) = cx.tcx.layout_of(cx.typing_env().as_query_input(ty))
+    {
+        layout.layout.size().bytes() == 0
+    } else {
+        false
+    }
+}
+
+// Make sure the raw pointer has no metadata
+fn is_narrow_raw_ptr<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    if let ty::RawPtr(_inner, _) = ty.kind() {
+        ty.pointee_metadata_ty_or_projection(cx.tcx).is_unit()
+    } else {
+        false
+    }
+}
+
+// Enum with some fixed representation and no data-carrying variants
+fn is_enum_repr_c<'tcx>(_cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    if let ty::Adt(adt_def, _args) = ty.kind()
+        && adt_def.is_enum()
+        && adt_def.repr().inhibit_struct_field_reordering()
+    {
+        adt_def.is_payloadfree()
+    } else {
+        false
+    }
+}
+
+// #[repr(transparent)] structures are also OK if the only non-zero
+// sized field contains a volatile-safe type
+fn is_struct_repr_transparent<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    if let ty::Adt(adt_def, args) = ty.kind()
+        && adt_def.is_struct()
+        && adt_def.repr().transparent()
+        && let [fieldty] = adt_def
+            .all_fields()
+            .filter_map(|field| {
+                let fty = field.ty(cx.tcx, args);
+                if is_zero_sized_ty(cx, fty) { None } else { Some(fty) }
+            })
+            .collect::<Vec<_>>()
+            .as_slice()
+    {
+        is_volatile_safe_ty(cx, *fieldty)
+    } else {
+        false
+    }
+}
+
+// SIMD can be useful to get larger atomic loads/stores, though this is still
+// pretty machine-dependent.
+fn is_simd_repr<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    if let ty::Adt(adt_def, _args) = ty.kind()
+        && adt_def.is_struct()
+        && adt_def.repr().simd()
+    {
+        let (_size, simdty) = ty.simd_size_and_type(cx.tcx);
+        is_volatile_safe_ty(cx, simdty)
+    } else {
+        false
+    }
+}
+
+// We can't know about a generic type, so just let it pass to avoid noise
+fn is_generic<'tcx>(_cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    ty.flags().intersects(ty::TypeFlags::HAS_PARAM)
```

Comment:
`ty.flags()` will be removed at some point. Can't you use `ty.has_param()`? Or `ty.has_non_region_param()` if a generic lifetime wouldn't change the lint?

Also, this function is so short that you could probably remove it and replace its call by its content.